### PR TITLE
Add a mock clipboard to interactive window tests

### PR DESCRIPTION
**Pull request by genlu**
Refactor `InteractiveWindow` and create a mock clipboard for interactive window tests to eliminate the COM exceptions we saw during CI build.

This should fix https://github.com/dotnet/roslyn/issues/5544. All related tests are re-enabled in this PR.

@cston @KevinH-MS @amcasey @ManishJayaswal @tmat 

FYI @jaredpar  @jmarolf

### CODE REVIEW

**Comment by jaredpar** [comment]
:+1:

**Comment by ManishJayaswal** [comment]
:+1:

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs [hunk 1]
```diff
@@ -1362,4 +1366,23 @@ internal static void Copy(this IInteractiveWindowOperations operations)
...
+        internal override bool ContainsData(string format) =>  _data?.GetData(format) != null;
+
+        internal override object GetData(string format) => _data?.GetData(format);
+
+        internal override bool ContainsText() => _data != null ? _data.ContainsText() : false;
+
+        internal override string GetText() => _data?.GetText();
+
+        internal override void SetDataObject(object data, bool copy) => _data = (DataObject)data;
+    }
```
Comment (cston) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Should _data be null initially and after Clear()?

**Comment by cston** [comment]
LGTM

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs (hunk 1 above)
Comment (genlu) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: @cston Yeah, I just had a closer look at the Clipboard API, SetData/Text/etc methods do expect a non-null argument, so I guess what you suggested would be a more accurate mock. Will fix. Thanks!

File: src/InteractiveWindow/Editor/InteractiveWindow.cs [hunk 2]
```diff
@@ -538,6 +540,32 @@ private static void DoEvents()
 
         internal event Action<State> StateChanged;
 
-#endregion
+        #endregion
+
+        private sealed class SystemWindowsClipboardWrapper : InteractiveWindowClipboard
```
Comment (tmat) [inline src/InteractiveWindow/Editor/InteractiveWindow.cs]: I'd just call it SystemClipboard

File: src/InteractiveWindow/Editor/InteractiveWindow.cs [hunk 3]
```diff
@@ -538,6 +540,32 @@ private static void DoEvents()
...
+
+            internal override bool ContainsText() => Clipboard.ContainsText();
+
+            internal override string GetText() => Clipboard.GetText();
+
+            internal override void SetDataObject(object data, bool copy) => Clipboard.SetDataObject(data, copy);
+        }
+    }
+
+    internal abstract class InteractiveWindowClipboard
```
Comment (tmat) [inline src/InteractiveWindow/Editor/InteractiveWindow.cs]: Move to a separate file.

File: src/InteractiveWindow/Editor/InteractiveWindow.cs (hunk 2 above)
Comment (tmat) [inline src/InteractiveWindow/Editor/InteractiveWindow.cs]: And move to a separate file.

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs [hunk 4]
```diff
@@ -20,11 +20,15 @@ public class InteractiveWindowTests : IDisposable
...
         private List<InteractiveWindow.State> _states;
+        private readonly TestClipboard _testClipboard; 
 
         public InteractiveWindowTests()
         {
             _states = new List<InteractiveWindow.State>();
             _testHost = new InteractiveWindowTestHost(_states.Add);
+            _testClipboard = new TestClipboard();
+            ((InteractiveWindow)Window).InteractiveWindowClipboard = _testClipboard;
+
```
Comment (tmat) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Extra line

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs [hunk 5]
```diff
@@ -1362,4 +1366,23 @@ internal static void Copy(this IInteractiveWindowOperations operations)
             ((IInteractiveWindowOperations2)operations).Copy();
         }
     }
+
+    internal sealed class TestClipboard : InteractiveWindowClipboard
```
Comment (tmat) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Move to separate file.

File: src/InteractiveWindow/Editor/InteractiveWindow.cs [hunk 6]
```diff
@@ -46,6 +46,8 @@ internal partial class InteractiveWindow : IInteractiveWindow, IInteractiveWindo
         /// </remarks>
         private readonly UIThreadOnly _uiOnly;
 
+        internal InteractiveWindowClipboard InteractiveWindowClipboard { get; set; } = new SystemClipboard();
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.cs]: Consider adding a comment that the setter is a test hook.

**Comment by amcasey** [comment]
:+1:

File: src/InteractiveWindow/Editor/InteractiveWindow.cs (hunk 6 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.cs]: @amcasey Will do with another PR. This is blocking my other work and the CI builds take forever to finish :)

**Pull request merged**